### PR TITLE
Issue #615: add zero boundary option to resampler and set as default

### LINKED ISSUE #615: 3d linear resampler should have "zero" as default extrapolation
## Subject of the issue

Describe your issue here.

Currently in https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/layer_util.py/resample_linear
the "replicate" is implemented, clip the coordinates outside of the region that has intensity values. This has a number of issues when registering a moving image that has smaller FOV to a fixed has a bigger FOV, making registration less overlap-invariant. This causes the problem for both images and labels.

An zero extrapolation can mitigate this to certain degree and should be set default. This should be treated as a bug-fix, rather than a feature.

## Steps to reproduce

The current paired_ct_lung trained model can be used as an example to reproduce the visible results

## Expected behaviour

- For bug-fix, change the extrapolation to zeros (background or negative class)
- for future, the extrapolation method can be made user-defined and the extrapolation value may be customisable too.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Add zero boundary option to resampler, if the location is on the boundary (corners or edges) the values will be zero.

Fixes #615

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
@QianyeYang @acasamitjana @s-sd would you have time to check the PR in details, really check the logics inside ^^

**Review by YipengHu — APPROVED** [review]

File: deepreg/model/layer_util.py [hunk 1]
```diff
@@ -114,36 +116,37 @@ def pyramid_combination(values: list, weights: list) -> tf.Tensor:
 
     The weights correspond to the floor corners.
     For example, when num_dimension = len(loc_shape) = num_bits = 3,
-    weights = [w1, w2, w3] (ignoring the batch dimension).
+    weight_floor = [w1, w2, w3] (ignoring the batch dimension).
```
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/layer_util.py:119]: Oh I are right, will do so

**Comment by mathpluscode** [comment]
Regarding the performance on the resampler, I will do a new ticket

**Review by acasamitjana — APPROVED** [review]
Great, thanks! Adding the nearest version shouldn't be complicated. I can help if you want.

**Review by s-sd — COMMENTED** [review]
Looks great @mathpluscode! I have added a comment for a suggestion. It looks like you wanted to make another change as well following a previous comment from @YipengHu so let me know once you're done with that or other comments and I'd be happy to approve.

**Comment by mathpluscode** [comment]
> Great, thanks! Adding the nearest version shouldn't be complicated. I can help if you want.

Sure! Can you create yet another ticket for this ^^

File: deepreg/model/layer_util.py [hunk 2]
```diff
@@ -114,36 +116,37 @@ def pyramid_combination(values: list, weights: list) -> tf.Tensor:
...
+          * ((1-y) * w2 + y * u2)
+          * ((1-z) * w3 + z * u3)
+
+    Let
 
-     = (W_xy * (1-z)) * w3 + (W_xy * z) * (1-w3)
+    W_xy = ((1-x) * w1 + x * u1)
+         * ((1-y) * w2 + y * u2)
 
-    where W_xy is the weight for (x, y), let
```
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/layer_util.py:133]: will do later today!

File: deepreg/model/layer_util.py [hunk 3]
```diff
@@ -156,37 +159,57 @@ def pyramid_combination(values: list, weights: list) -> tf.Tensor:
...
+            "num_dim = len(weight_floor), "
             "len(values) must be 2 ** num_dim, "
-            "But len(weights) = {}, len(values) = {}".format(len(weights), len(values))
+            f"But len(weight_floor) = {len(weight_floor)}, len(values) = {len(values)}"
         )
 
-    if len(weights) == 1:  # one dimension
-        return values[0] * weights[0] + values[1] * (1 - weights[0])
+    if len(weight_floor) == 1:  # one dimension
+        return values[0] * weight_floor[0] + values[1] * weight_ceil[0]
```
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/layer_util.py:187]: will do later today!

File: deepreg/model/layer_util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/layer_util.py:133]: done

File: deepreg/model/layer_util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/layer_util.py:119]: done

File: deepreg/model/layer_util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/layer_util.py:187]: added assert weight floor should == weight ceil

**Pull request merged**